### ISSUE TITLE
panic: count of short block: got 0, exp 1

### ISSUE THREAD

**Issue by weshmashian** [influxdata/influxdb#7897]
__System info:__
InfluxDB version 1.2.0, branch master, commit b7bb7e8359642b6e071735b50ae41f5eb343fd42

__Steps to reproduce:__

1. Start up InfluxDB and wait.

__Expected behavior:__ 
InfluxDB works normally.

__Actual behavior:__ 

InfluxDB starts compacting databases and panics with the following trace:

```
panic: count of short block: got 0, exp 1

goroutine 53530 [running]:
panic(0x9d12c0, 0xc6ad599740)
        /usr/local/go/src/runtime/panic.go:500 +0x1a1
github.com/influxdata/influxdb/tsdb/engine/tsm1.BlockCount(0x0, 0x0, 0x0, 0x8)
        /root/go/src/github.com/influxdata/influxdb/tsdb/engine/tsm1/encoding.go:224 +0x315
github.com/influxdata/influxdb/tsdb/engine/tsm1.(*tsmKeyIterator).combine(0xc4a7675b00, 0x0, 0xa7a980, 0x1, 0xc7086c8c00)
        /root/go/src/github.com/influxdata/influxdb/tsdb/engine/tsm1/compact.go:1165 +0x6eb
github.com/influxdata/influxdb/tsdb/engine/tsm1.(*tsmKeyIterator).merge(0xc4a7675b00)
        /root/go/src/github.com/influxdata/influxdb/tsdb/engine/tsm1/compact.go:1101 +0x100
github.com/influxdata/influxdb/tsdb/engine/tsm1.(*tsmKeyIterator).Next(0xc4a7675b00, 0xc613bb36ea)
        /root/go/src/github.com/influxdata/influxdb/tsdb/engine/tsm1/compact.go:1068 +0x2bd
github.com/influxdata/influxdb/tsdb/engine/tsm1.(*Compactor).write(0xc5183ba050, 0xc57fbdbdb0, 0x47, 0xe30720, 0xc4a7675b00, 0x0, 0x0)
        /root/go/src/github.com/influxdata/influxdb/tsdb/engine/tsm1/compact.go:774 +0x29b
github.com/influxdata/influxdb/tsdb/engine/tsm1.(*Compactor).writeNewFiles(0xc5183ba050, 0x22f, 0x6, 0xe30720, 0xc4a7675b00, 0xe30720, 0xc4a7675b00, 0x0, 0x0, 0xc57cba1bf8)
        /root/go/src/github.com/influxdata/influxdb/tsdb/engine/tsm1/compact.go:728 +0x284
github.com/influxdata/influxdb/tsdb/engine/tsm1.(*Compactor).compact(0xc5183ba050, 0xbacc00, 0xc570c61aa0, 0x2, 0x2, 0x0, 0x0, 0x0, 0x0, 0x0)
        /root/go/src/github.com/influxdata/influxdb/tsdb/engine/tsm1/compact.go:654 +0x52a
github.com/influxdata/influxdb/tsdb/engine/tsm1.(*Compactor).CompactFull(0xc5183ba050, 0xc570c61aa0, 0x2, 0x2, 0x0, 0x0, 0x0, 0x0, 0x0)
        /root/go/src/github.com/influxdata/influxdb/tsdb/engine/tsm1/compact.go:672 +0x18e
github.com/influxdata/influxdb/tsdb/engine/tsm1.(*compactionStrategy).compactGroup.func1(0xc56eba0930, 0xc570c61aa0, 0x2, 0x2, 0x0, 0x0, 0x0, 0x0, 0x0)
        /root/go/src/github.com/influxdata/influxdb/tsdb/engine/tsm1/engine.go:1072 +0x15d
github.com/influxdata/influxdb/tsdb/engine/tsm1.(*compactionStrategy).compactGroup(0xc56eba0930, 0x0)
        /root/go/src/github.com/influxdata/influxdb/tsdb/engine/tsm1/engine.go:1074 +0x47d
github.com/influxdata/influxdb/tsdb/engine/tsm1.(*compactionStrategy).Apply.func1(0xc57cac3380, 0xc56eba0930, 0x0)
        /root/go/src/github.com/influxdata/influxdb/tsdb/engine/tsm1/engine.go:1047 +0x5d
created by github.com/influxdata/influxdb/tsdb/engine/tsm1.(*compactionStrategy).Apply
        /root/go/src/github.com/influxdata/influxdb/tsdb/engine/tsm1/engine.go:1048 +0xcc
[I] 2017-01-27T08:27:13Z InfluxDB starting, version 1.2.0, branch master, commit b7bb7e8359642b6e071735b50ae41f5eb343fd42
```

__Additional info:__ 

We have ~250 active databases and users running under one Influx instance so this might very well be the source of our issues.

The above panic was (for now) mitigated by syncing known good datadir and downgrading InfluxDB to 1.2.0-rc1. Given that we're having issues with stability (#7640) we might not be able to try and reproduce this until we get some stable environment up and running.

**Comment by jwilder** [comment]
Can you elaborate on your hardware setup?  
* Is InfluxDB running on bare hardware, in a VM, container, etc?  
* Are the disks SSD and directly attached?  
* Is the datadir shared with any other processes? 
* Is there any storage management in the mix (RAID, NAS, SAN, etc..)?  
* Can you explain what you mean by "syncing known good datadir"?

**Comment by weshmashian** [comment]
We are running 3 instances of Influx, each one on it's own dedicated server. Out of the three, only one Influx is not crashing or showing any problems and we're using it as a source to reseed other two instances, referred here as a 'good datadir'.

Each server uses SSDs in RAID10 for data storage, and datadir is used exclusively by InfluxDB. In our case nothing is virtualized or shared, and everything is running on bare HW.

**Comment by weshmashian** [comment]
Running any DB in production without RAID is simply not an option in our case, as we need to ensure redundancy on several levels, despite having backups.

On a side note, I'm having a hard time understanding why would any RAID level cause any issues?

**Comment by timhallinflux** [comment]
@weshmashian -- are you still experiencing this issue? or shall we close this out?

**Comment by weshmashian** [comment]
We've managed to work around this. IIRC the main issue was insane number of points Docker metrics were sending. It's been working ever since we stopped doing that and it survived the Scream Test.

**Comment by dgnorton** [comment]
Fixed in `v1.8.0` by https://github.com/influxdata/influxdb/pull/17570
